File: modules/cadence-audio/android/src/main/java/expo/modules/cadenceaudio/CadenceAudioModule.kt
package expo.modules.cadenceaudio

import android.content.Context
import android.media.AudioAttributes
import android.media.AudioFocusRequest
import android.media.AudioFormat
import android.media.AudioManager
import android.media.AudioTrack
import android.os.Build
import android.os.Handler
import android.os.Looper
import android.os.Process
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import kotlin.concurrent.thread
import kotlin.math.PI
import kotlin.math.abs
import kotlin.math.exp
import kotlin.math.max
import kotlin.math.min
import kotlin.math.sin
import kotlin.random.Random

// Order matters: index === position in this list.
private val SOUND_IDS = listOf("beep", "woodfish", "click")
private fun soundIndex(id: String): Int = SOUND_IDS.indexOf(id).let { if (it < 0) 1 else it }
private fun clampBpm(b: Double): Double = min(250.0, max(100.0, Math.round(b).toDouble()))

/**
 * Sample-accurate cadence engine on AudioTrack (ENCODING_PCM_FLOAT, streaming).
 *
 * A high-priority render thread fills fixed PCM blocks, placing each click at an
 * exact sample offset via a `samplesUntilNextBeat` counter that re-reads the live
 * interval at every beat — zero accumulating drift, sub-sample jitter. The blocking
 * `AudioTrack.write` paces the loop to the hardware clock.
 *
 * We never request audio focus, so beats mix on top of any music/podcast/video
 * (PRD §3.2 coexist). USAGE_ASSISTANCE_SONIFICATION marks them as cues so the
 * system won't arbitrarily duck them.
 */
class CadenceAudioModule : Module() {
  // ── Control state (written by JS thread, read by render thread) ──────────
  @Volatile private var intervalSamples = 16000
  @Volatile private var volume = 0.72f
  @Volatile private var running = false
  @Volatile private var selectedSound = 1 // woodfish
  @Volatile private var beatCounter = 0
  @Volatile private var mixWithOthers = true
  @Volatile private var ducking = false

  // Audio focus is only held while ducking; otherwise we deliberately request
  // none so beats mix on top of other audio (PRD §3.2 coexist).
  private var focusRequest: AudioFocusRequest? = null

  private var sampleRate = 48000
  private var track: AudioTrack? = null
  private var renderThread: Thread? = null
  @Volatile private var threadAlive = false
  private var prepared = false

  private var clickBuffers: Array<FloatArray> = arrayOf()

  // ── Render-thread-owned state ────────────────────────────────────────────
  @Volatile private var samplesUntilNextBeat = 0
  private val voiceCount = 8
  private val voiceActive = BooleanArray(voiceCount)
  private val voiceSound = IntArray(voiceCount)
  private val voicePos = IntArray(voiceCount)

  private val mainHandler = Handler(Looper.getMainLooper())
  private var lastEmittedBeat = 0
  @Volatile private var polling = false

  override fun definition() = ModuleDefinition {
    Name("CadenceAudio")

    Events("onBeat")

    Property("isRunning") { running }

    Function("prepare") { mix: Boolean -> prepare(mix) }

    Function("start") { bpm: Double ->
      intervalSamples = intervalSamplesFor(bpm)
      if (!prepared) prepare(mixWithOthers)
      samplesUntilNextBeat = 0 // first beat fires immediately
      beatCounter = 0
      lastEmittedBeat = 0
      running = true
      ensurePlaying()
    }

    Function("stop") { running = false }

    Function("setBpm") { bpm: Double -> intervalSamples = intervalSamplesFor(bpm) }

    Function("setVolume") { v: Double -> volume = min(1.0, max(0.0, v)).toFloat() }

    Function("setSound") { id: String -> selectedSound = soundIndex(id) }

    Function("setMixWithOthers") { mix: Boolean -> mixWithOthers = mix }

    Function("setDucking") { duck: Boolean ->
      ducking = duck
      if (duck) requestDuckFocus() else abandonFocus()
    }

    OnDestroy { teardown() }
  }

  private fun intervalSamplesFor(bpm: Double): Int {
    val b = clampBpm(bpm)
    return max(1, Math.round(sampleRate * 60.0 / b).toInt())
  }

  // MARK: - Setup

  private fun prepare(mix: Boolean) {
    mixWithOthers = mix
    if (prepared) return

    val native = AudioTrack.getNativeOutputSampleRate(AudioManager.STREAM_MUSIC)
    sampleRate = if (native > 0) native else 48000
    synthesize()

    val channelMask = AudioFormat.CHANNEL_OUT_STEREO
    val minBuf = AudioTrack.getMinBufferSize(sampleRate, channelMask, AudioFormat.ENCODING_PCM_FLOAT)
    val bufBytes = max(minBuf, 4 /*bytes*/ * 2 /*ch*/ * 1024)

    // USAGE_MEDIA routes to the music stream (the volume the user controls with
    // the volume keys), so beats are reliably audible. We still never request
    // audio focus, so they mix on top of any playing music (PRD §3.2 coexist).
    val attrs = AudioAttributes.Builder()
      .setUsage(AudioAttributes.USAGE_MEDIA)
      .setContentType(AudioAttributes.CONTENT_TYPE_MUSIC)
      .build()
    val fmt = AudioFormat.Builder()
      .setSampleRate(sampleRate)
      .setEncoding(AudioFormat.ENCODING_PCM_FLOAT)
      .setChannelMask(channelMask)
      .build()

    val builder = AudioTrack.Builder()
      .setAudioAttributes(attrs)
      .setAudioFormat(fmt)
      .setBufferSizeInBytes(bufBytes)
      .setTransferMode(AudioTrack.MODE_STREAM)
    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
      builder.setPerformanceMode(AudioTrack.PERFORMANCE_MODE_LOW_LATENCY)
    }
    track = builder.build()

    startBeatPoll()
    prepared = true
  }

  private fun ensurePlaying() {
    val t = track ?: return
    if (t.playState != AudioTrack.PLAYSTATE_PLAYING) t.play()
    if (renderThread?.isAlive != true) {
      threadAlive = true
      renderThread = thread(name = "cadence-render", priority = Thread.MAX_PRIORITY) { renderLoop() }
    }
  }

  // MARK: - Render loop

  private fun renderLoop() {
    Process.setThreadPriority(Process.THREAD_PRIORITY_URGENT_AUDIO)
    val t = track ?: return
    val blockFrames = 256
    val channels = 2
    val out = FloatArray(blockFrames * channels)

    while (threadAlive) {
      val interval = intervalSamples
      val vol = volume
      val isRun = running

      for (i in 0 until blockFrames) {
        var s = 0f

        for (v in 0 until voiceCount) {
          if (voiceActive[v]) {
            val buf = clickBuffers[voiceSound[v]]
            val pos = voicePos[v]
            s += buf[pos]
            val next = pos + 1
            if (next >= buf.size) voiceActive[v] = false else voicePos[v] = next
          }
        }

        if (isRun) {
          samplesUntilNextBeat -= 1
          if (samplesUntilNextBeat <= 0) {
            val si = selectedSound
            for (v in 0 until voiceCount) {
              if (!voiceActive[v]) {
                voiceActive[v] = true
                voiceSound[v] = si
                voicePos[v] = 0
                break
              }
            }
            samplesUntilNextBeat += max(1, interval) // re-rate at the boundary
            beatCounter += 1
          }
        }

        s *= vol
        val o = i * channels
        out[o] = s
        out[o + 1] = s
      }

      val written = t.write(out, 0, out.size, AudioTrack.WRITE_BLOCKING)
      if (written < 0) break
    }
  }

  // MARK: - Click synthesis (short, normalized, zero-silence edges)

  private fun synthesize() {
    if (clickBuffers.isNotEmpty()) return
    // freq, decay(s), dur(s), noise
    val grains = arrayOf(
      doubleArrayOf(1900.0, 0.012, 0.035, 0.0),  // beep
      doubleArrayOf(720.0, 0.018, 0.055, 0.06),  // woodfish
      doubleArrayOf(3000.0, 0.004, 0.014, 0.15), // click
    )
    clickBuffers = Array(grains.size) { synth(grains[it][0], grains[it][1], grains[it][2], grains[it][3]) }
  }

  private fun synth(freq: Double, decay: Double, dur: Double, noise: Double): FloatArray {
    val len = max(1, (dur * sampleRate).toInt())
    val fadeOut = max(1, (0.002 * sampleRate).toInt())
    val data = FloatArray(len)
    var peak = 0f
    for (i in 0 until len) {
      val t = i.toDouble() / sampleRate
      val env = exp(-t / decay)
      val tone = sin(2.0 * PI * freq * t)
      val n = if (noise > 0) Random.nextDouble(-1.0, 1.0) * noise else 0.0
      var s = ((tone + n) * env).toFloat()
      if (i < 32) s *= i / 32f                 // sub-ms attack from zero
      if (i > len - fadeOut) s *= (len - i).toFloat() / fadeOut
      data[i] = s
      if (abs(s) > peak) peak = abs(s)
    }
    if (peak > 0) {
      val norm = 0.9f / peak
      for (i in 0 until len) data[i] *= norm
    }
    return data
  }

  // MARK: - Beat events (off the render thread)

  private fun startBeatPoll() {
    if (polling) return
    polling = true
    val poll = object : Runnable {
      override fun run() {
        if (!polling) return
        val c = beatCounter
        if (c != lastEmittedBeat) {
          lastEmittedBeat = c
          sendEvent("onBeat", mapOf("beatIndex" to c))
        }
        mainHandler.postDelayed(this, 30)
      }
    }
    mainHandler.post(poll)
  }

  // MARK: - Audio focus (only held while ducking)

  private fun audioManager(): AudioManager? =
    appContext.reactContext?.getSystemService(Context.AUDIO_SERVICE) as? AudioManager

  private fun requestDuckFocus() {
    val am = audioManager() ?: return
    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
      if (focusRequest != null) return
      val req = AudioFocusRequest.Builder(AudioManager.AUDIOFOCUS_GAIN_TRANSIENT_MAY_DUCK)
        .setAudioAttributes(
          AudioAttributes.Builder()
            .setUsage(AudioAttributes.USAGE_MEDIA)
            .setContentType(AudioAttributes.CONTENT_TYPE_MUSIC)
            .build()
        )
        // We synthesize beats ourselves and never pause; ignore focus changes.
        .setWillPauseWhenDucked(false)
        .build()
      focusRequest = req
      am.requestAudioFocus(req)
    } else {
      @Suppress("DEPRECATION")
      am.requestAudioFocus(
        null,
        AudioManager.STREAM_MUSIC,
        AudioManager.AUDIOFOCUS_GAIN_TRANSIENT_MAY_DUCK
      )
    }
  }

  private fun abandonFocus() {
    val am = audioManager() ?: return
    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
      focusRequest?.let { am.abandonAudioFocusRequest(it) }
      focusRequest = null
    } else {
      @Suppress("DEPRECATION")
      am.abandonAudioFocus(null)
    }
  }

  // MARK: - Teardown

  private fun teardown() {
    running = false
    threadAlive = false
    polling = false
    abandonFocus()
    try {
      renderThread?.join(200)
    } catch (_: InterruptedException) {
    }
    renderThread = null
    track?.let {
      try {
        it.stop()
      } catch (_: Exception) {
      }
      it.release()
    }
    track = null
    prepared = false
  }
}

File: modules/cadence-audio/ios/CadenceAudioModule.swift
import ExpoModulesCore
import AVFoundation

// Order matters: index === position in this array.
private let SOUND_IDS = ["beep", "woodfish", "click"]

private func soundIndex(_ id: String) -> Int {
  return SOUND_IDS.firstIndex(of: id) ?? 1
}

private func clampBpm(_ bpm: Double) -> Double {
  return min(250.0, max(100.0, bpm.rounded()))
}

/// Realtime state shared between the control thread (JS calls) and the audio
/// render thread. All hot fields are word-size scalars; on arm64 their reads
/// and writes are atomic enough for a metronome (a torn BPM is impossible for
/// an aligned Int). Beat *placement* is counted in samples, so there is no
/// accumulating-timer drift regardless of when control writes land.
private final class RTState {
  var intervalSamples: Int = 16000
  var volume: Float = 0.72
  var running: Bool = false
  var selectedSound: Int = 1 // woodfish

  // Owned by the render thread only.
  var samplesUntilNextBeat: Int = 0
  var beatCounter: Int = 0

  // Synthesized click buffers (raw pointers → no ARC in the render loop).
  var bufPtrs: [UnsafeMutablePointer<Float>] = []
  var bufLens: [Int] = []

  // Fixed voice pool (clicks never overlap at ≤250 BPM, but be safe).
  static let voiceCount = 8
  var voiceActive = [Bool](repeating: false, count: voiceCount)
  var voiceSound = [Int](repeating: 0, count: voiceCount)
  var voicePos = [Int](repeating: 0, count: voiceCount)
}

public class CadenceAudioModule: Module {
  private let engine = AVAudioEngine()
  private var sourceNode: AVAudioSourceNode?
  private let rt = RTState()
  private var sampleRate: Double = 48000
  private var prepared = false
  private var mixWithOthers = true
  private var ducking = false

  private var beatTimer: DispatchSourceTimer?
  private var lastEmittedBeat = 0
  private var wasRunningBeforeInterruption = false

  public func definition() -> ModuleDefinition {
    Name("CadenceAudio")

    Events("onBeat")

    Property("isRunning") { [weak self] () -> Bool in
      return self?.rt.running ?? false
    }

    Function("prepare") { (mix: Bool) in
      self.prepare(mix: mix)
    }

    Function("start") { (bpm: Double) in
      self.rt.intervalSamples = self.intervalSamplesFor(bpm)
      if !self.prepared { self.prepare(mix: self.mixWithOthers) }
      self.rt.samplesUntilNextBeat = 0 // fire the first beat immediately
      self.rt.beatCounter = 0
      self.lastEmittedBeat = 0
      self.rt.running = true
      self.ensureEngineRunning()
    }

    Function("stop") {
      self.rt.running = false
    }

    Function("setBpm") { (bpm: Double) in
      self.rt.intervalSamples = self.intervalSamplesFor(bpm)
    }

    Function("setVolume") { (v: Double) in
      self.rt.volume = Float(min(1.0, max(0.0, v)))
    }

    Function("setSound") { (id: String) in
      self.rt.selectedSound = soundIndex(id)
    }

    Function("setMixWithOthers") { (mix: Bool) in
      self.mixWithOthers = mix
      try? self.configureSession(mix: mix)
    }

    Function("setDucking") { (duck: Bool) in
      self.ducking = duck
      try? self.configureSession(mix: self.mixWithOthers)
    }

    OnDestroy {
      self.teardown()
    }
  }

  // MARK: - Setup

  private func intervalSamplesFor(_ bpm: Double) -> Int {
    let b = clampBpm(bpm)
    return max(1, Int((sampleRate * 60.0 / b).rounded()))
  }

  private func configureSession(mix: Bool) throws {
    let session = AVAudioSession.sharedInstance()
    var options: AVAudioSession.CategoryOptions = []
    if mix {
      // `.duckOthers` implies mixing, and lowers (rather than stops) other audio.
      options = ducking ? [.duckOthers] : [.mixWithOthers]
    }
    try session.setCategory(.playback, mode: .default, options: options)
    try session.setActive(true)
    sampleRate = session.sampleRate
  }

  private func prepare(mix: Bool) {
    if prepared {
      try? configureSession(mix: mix)
      return
    }
    mixWithOthers = mix
    do {
      try configureSession(mix: mix)
    } catch {
      NSLog("CadenceAudio: session config failed: \(error)")
    }

    synthesizeClicks()

    let format = AVAudioFormat(standardFormatWithSampleRate: sampleRate, channels: 2)!
    let rt = self.rt // local; captured unowned below to avoid ARC in the render loop
    let node = AVAudioSourceNode(format: format) { [unowned rt] _, _, frameCount, ablPointer in
      let abl = UnsafeMutableAudioBufferListPointer(ablPointer)
      let frames = Int(frameCount)
      let channels = abl.count
      let vol = rt.volume
      let interval = rt.intervalSamples
      let running = rt.running

      for frame in 0..<frames {
        var s: Float = 0

        // Sum and advance active voices.
        for v in 0..<RTState.voiceCount where rt.voiceActive[v] {
          let si = rt.voiceSound[v]
          let pos = rt.voicePos[v]
          s += rt.bufPtrs[si][pos]
          let next = pos + 1
          if next >= rt.bufLens[si] {
            rt.voiceActive[v] = false
          } else {
            rt.voicePos[v] = next
          }
        }

        // Beat boundary — sample-exact.
        if running {
          rt.samplesUntilNextBeat -= 1
          if rt.samplesUntilNextBeat <= 0 {
            let si = rt.selectedSound
            for v in 0..<RTState.voiceCount where !rt.voiceActive[v] {
              rt.voiceActive[v] = true
              rt.voiceSound[v] = si
              rt.voicePos[v] = 0
              break
            }
            rt.samplesUntilNextBeat += max(1, interval) // re-rate at the boundary
            rt.beatCounter &+= 1
          }
        }

        s *= vol
        for ch in 0..<channels {
          if let mdata = abl[ch].mData {
            mdata.assumingMemoryBound(to: Float.self)[frame] = s
          }
        }
      }
      return noErr
    }

    sourceNode = node
    engine.attach(node)
    engine.connect(node, to: engine.mainMixerNode, format: format)
    engine.prepare()

    registerInterruptionObserver()
    startBeatTimer()
    prepared = true
  }

  private func ensureEngineRunning() {
    guard prepared, !engine.isRunning else { return }
    do {
      try AVAudioSession.sharedInstance().setActive(true)
      try engine.start()
    } catch {
      NSLog("CadenceAudio: engine start failed: \(error)")
    }
  }

  // MARK: - Click synthesis

  private func synthesizeClicks() {
    guard rt.bufPtrs.isEmpty else { return }
    // freq, decay(s), dur(s), noise — short, normalized, zero-silence edges.
    let grains: [(Double, Double, Double, Double)] = [
      (1900, 0.012, 0.035, 0.0),  // beep
      (720, 0.018, 0.055, 0.06),  // woodfish
      (3000, 0.004, 0.014, 0.15), // click
    ]
    for g in grains {
      let arr = synth(freq: g.0, decay: g.1, dur: g.2, noise: g.3, sr: sampleRate)
      let ptr = UnsafeMutablePointer<Float>.allocate(capacity: arr.count)
      ptr.initialize(from: arr, count: arr.count)
      rt.bufPtrs.append(ptr)
      rt.bufLens.append(arr.count)
    }
  }

  private func synth(freq: Double, decay: Double, dur: Double, noise: Double, sr: Double) -> [Float] {
    let len = max(1, Int(dur * sr))
    let fadeOut = max(1, Int(0.002 * sr))
    var data = [Float](repeating: 0, count: len)
    var peak: Float = 0
    for i in 0..<len {
      let t = Double(i) / sr
      let env = exp(-t / decay)
      let tone = sin(2.0 * Double.pi * freq * t)
      let n = noise > 0 ? Double.random(in: -1...1) * noise : 0
      var s = Float((tone + n) * env)
      if i < 32 { s *= Float(i) / 32.0 } // sub-ms attack from zero
      if i > len - fadeOut { s *= Float(len - i) / Float(fadeOut) }
      data[i] = s
      peak = max(peak, abs(s))
    }
    if peak > 0 {
      let norm = 0.9 / peak
      for i in 0..<len { data[i] *= norm }
    }
    return data
  }

  // MARK: - Beat events (off the realtime thread)

  private func startBeatTimer() {
    let timer = DispatchSource.makeTimerSource(queue: DispatchQueue.global(qos: .userInitiated))
    timer.schedule(deadline: .now(), repeating: .milliseconds(30))
    timer.setEventHandler { [weak self] in
      guard let self = self else { return }
      let c = self.rt.beatCounter
      if c != self.lastEmittedBeat {
        self.lastEmittedBeat = c
        self.sendEvent("onBeat", ["beatIndex": c])
      }
    }
    timer.resume()
    beatTimer = timer
  }

  // MARK: - Interruptions (PRD §4.1 — real interruptions pause)

  private func registerInterruptionObserver() {
    NotificationCenter.default.addObserver(
      self,
      selector: #selector(handleInterruption(_:)),
      name: AVAudioSession.interruptionNotification,
      object: AVAudioSession.sharedInstance()
    )
  }

  @objc private func handleInterruption(_ note: Notification) {
    guard let info = note.userInfo,
          let raw = info[AVAudioSessionInterruptionTypeKey] as? UInt,
          let type = AVAudioSession.InterruptionType(rawValue: raw) else { return }
    switch type {
    case .began:
      wasRunningBeforeInterruption = rt.running
      rt.running = false
      engine.pause()
    case .ended:
      let shouldResume: Bool
      if let optRaw = info[AVAudioSessionInterruptionOptionKey] as? UInt {
        shouldResume = AVAudioSession.InterruptionOptions(rawValue: optRaw).contains(.shouldResume)
      } else {
        shouldResume = false
      }
      if shouldResume && wasRunningBeforeInterruption {
        ensureEngineRunning()
        rt.running = true
      }
    @unknown default:
      break
    }
  }

  // MARK: - Teardown

  private func teardown() {
    beatTimer?.cancel()
    beatTimer = nil
    NotificationCenter.default.removeObserver(self)
    if engine.isRunning { engine.stop() }
    for ptr in rt.bufPtrs { ptr.deallocate() }
    rt.bufPtrs.removeAll()
    rt.bufLens.removeAll()
    try? AVAudioSession.sharedInstance().setActive(false, options: .notifyOthersOnDeactivation)
  }
}
